#Importing Libraries


In [3]:
#import libraries
import pandas as pd
import numpy as np
import os

# Importing Datasets

In [5]:
df_prods=pd.read_csv(r'C:\Users\dsubr\Downloads\11-04-2025 Instacart Basket Analysis\02. Data\Original Data\products.csv',index_col=False)

In [7]:
df_ords=pd.read_csv(r'C:\Users\dsubr\Downloads\11-04-2025 Instacart Basket Analysis\02. Data\Prepared Data\orders_wrangled.csv',index_col=False)

# Create a dataframe

In [11]:
df_test=pd.DataFrame()

#Create a mixed type column

In [20]:
df_test['mix']=['a',12,'$',True]

In [22]:
df_test.head()

mix
0     a
1    12
2     $
3  True

# Data Consistency Checks

# Checking for mixed datatype

In [13]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [24]:
df_test['mix']=df_test['mix'].astype('str')

# Finding missing values

In [26]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [28]:
df_nan=df_prods[df_prods['product_name'].isnull() == True]

In [30]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing missing values

In [32]:
df_prods.shape

(49693, 5)

In [34]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [36]:
df_prods_clean.shape

(49677, 5)

In [38]:
df_prods.dropna(inplace=True)

# Finding duplicates

In [40]:
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [42]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [44]:
df_prods_clean.shape

(49677, 5)

In [46]:
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

# Task 4.5 

In [48]:
df_prods_clean_no_dups.shape

(49672, 5)

In [50]:
df_prods=df_prods_clean_no_dups

In [52]:
df_prods.shape

(49672, 5)

2.Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [51]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [57]:
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

3.Check for mixed-type data in your df_ords dataframe.

In [88]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\dsubr\AppData\Local\Temp\ipykernel_15184\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\dsubr\AppData\Local\Temp\ipykernel_15184\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\dsubr\AppData\Local\Temp\ipykernel_15184\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\dsubr\AppData\Local\Temp\ipykernel_15184\1946670780.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\dsubr\AppData\Local

4.If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

In [92]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


In [98]:
df_ords['order_id'].dtype

dtype('int64')

In [102]:
df_ords['order_id'] = df_ords['order_id'].astype('int32')

In [108]:
df_ords['user_id']=df_ords['user_id'].astype('int32')

In [112]:
df_ords['order_number']=df_ords['order_number'].astype('int32')

In [114]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [118]:
df_ords['orders_day_of_week']=df_ords['orders_day_of_week'].astype('int32')

In [120]:
df_ords.rename(columns = {'order_hour_of_day' : 'order_hour'}, inplace = True)

In [122]:
df_ords['order_hour']=df_ords['order_hour'].astype('int32')

In [124]:
df_ords['order_id'].dtype

dtype('int32')

In [126]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

         order_hour  days_since_prior_order  
count  3.421083e+06            3.214874e+06  
mean   1.345202e+01            1.111484e+01  
std    4.226088e+00            9.206737e+00  
min    0.000000e+00            0.000000e+00  
25%    1.000000e+01            4.000000e+00  
50%    1.300000e+01            7.000000e+00  
75%    1.600000e+01            1.500000e+01  
max    2.300000e+01            3.000000e+01

No mixed type values found.

5.Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [130]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour                     0
days_since_prior_order    206209
dtype: int64

In [136]:
df_ords.drop(columns=['eval_set'])

order_id  user_id  order_number  orders_day_of_week  order_hour  \
0         2539329        1             1                   2           8   
1         2398795        1             2                   3           7   
2          473747        1             3                   3          12   
3         2254736        1             4                   4           7   
4          431534        1             5                   4          15   
...           ...      ...           ...                 ...         ...   
3421078   2266710   206209            10                   5          18   
3421079   1854736   206209            11                   4          10   
3421080    626363   206209            12                   1          12   
3421081   2977660   206209            13                   1          12   
3421082    272231   206209            14                   6          14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [150]:
df_ords

order_id  user_id  order_number  orders_day_of_week  order_hour  \
0         2539329        1             1                   2           8   
1         2398795        1             2                   3           7   
2          473747        1             3                   3          12   
3         2254736        1             4                   4           7   
4          431534        1             5                   4          15   
...           ...      ...           ...                 ...         ...   
3421078   2266710   206209            10                   5          18   
3421079   1854736   206209            11                   4          10   
3421080    626363   206209            12                   1          12   
3421081   2977660   206209            13                   1          12   
3421082    272231   206209            14                   6          14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

6.Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

There are missing values in days_since_prior_order field.Based on the value 206209, retrieving the count based on userid.

df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

7.Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.


In [166]:
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour, days_since_prior_order]
Index: []

8.Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

df_ords 

df_ords has no duplicate values to be removed.

9.Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [54]:
path=r'C:\Users\dsubr\Downloads\11-04-2025 Instacart Basket Analysis\02. Data\Original Data\products.csv'

In [176]:
path

'C:\\Users\\dsubr\\Downloads\\11-04-2025 Instacart Basket Analysis\\02. Data\\Original Data\\products.csv'

In [56]:
df_prods.to_csv(r'C:\Users\dsubr\Downloads\11-04-2025 Instacart Basket Analysis\02. Data\Prepared Data\products_cleaned.csv',index=False)

In [58]:
df_ords.to_csv(r'C:\Users\dsubr\Downloads\11-04-2025 Instacart Basket Analysis\02. Data\Prepared Data\orders_cleaned.csv',index=False)